completion_aggregator/transformers.py

```diff
@@ -79,3 +84,119 @@ def transform(self, usage_info, block_structure):  # pylint: disable=unused-argu
             if completion_mode != XBlockCompletionMode.EXCLUDED:
                 aggregators = self.calculate_aggregators(block_structure, block_key)
                 block_structure.set_transformer_block_field(block_key, self, self.AGGREGATORS, aggregators)
+
+
+class BaseAggregatorXApiTransformer(XApiTransformer):
+    """
+    Base transformer for all completion aggregator events.
+    """
+
+    object_type = None
+
+    def get_object(self) -> Activity:
+        """
+        Get object for xAPI transformed event.
+        """
+        if not self.object_type:
+            raise NotImplementedError()  # pragma: no cover
+
+        return Activity(
+            id=self.get_object_iri("xblock", self.get_data("data.block_id")),
+            definition=ActivityDefinition(
+                type=self.object_type,
+            ),
+        )
+
+
+class BaseProgressTransformer(BaseAggregatorXApiTransformer):
+    """
+    Base transformer for completion aggregator progress events.
+    """
+
+    _verb = Verb(
+        id=constants.XAPI_VERB_PROGRESSED,
+        display=LanguageMap({constants.EN: constants.PROGRESSED}),
+    )
+    object_type = None
+    additional_fields = ('result', )
+
+    @openedx_filter(
+        filter_type="completion_aggregator.xapi.base_progress.get_object",
+    )
+    def get_object(self) -> Activity:
+        """
+        Get object for xAPI transformed event.
+        """
+        return super().get_object()
+
+    def get_result(self) -> Result:
+        """
+        Get result for xAPI transformed event.
+        """
+        return Result(
+            completion=self.get_data("data.percent") == 1.0,
```

Comment:
Can it ever be `True`? The code from https://github.com/open-craft/openedx-completion-aggregator/pull/206 already distinguishes `progress` and `completion` events by checking if the `percent < 1`.

If we want to keep it, would it make sense to call it "completed" instead, since it's a boolean?

Reply:
Unfortunately we don't have a choice about what these attributes are called -- tincan defines a fixed set of attributes for the [Result class](http://rusticisoftware.github.io/TinCanJS/doc/api/latest/classes/TinCan.Result.html).

But let me know if you'd rather I hard-coded `completion=False` for progress events, if this is confusing.